OSDOCS-19540: adds gRPC support RHCL (#111522)
openshift/openshift-docs: modules/proc-migrate-policies-to-grpc.adoc

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: PROCEDURE
[id="con-migrate-policies-to-grpc_{context}"]
= Migrating policies to `GRPCRoute` support

:FeatureName: `GRPCRoute` policy attachment support

include::snippets/technology-preview.adoc[]

[role="_abstract"]
When you want to replace your `HTTPRoute` custom resources (CRs) with `GRPCRoute` CRs, you must update your policies to reference a `GRPCRoute` CR, and configure other parameters in those policies as required for your use case.

[IMPORTANT]
====
A `GRPCRoute` CR cannot coexist with an `HTTPRoute` CR on the same hostname. If both gRPC and HTTP traffic must share a hostname, use an `HTTPRoute` CR. Otherwise, use separate hostnames for HTTP and gRPC traffic.
====

.Prerequisites

* You installed {ocp}.
* You installed {oc-first}.
* You installed the {mcpg} Operator.
* You created a Gateway object.

> * You created a `GRPCRoute` CR.

Reviewer: ```suggestion
* You have an existing HTTPRoute CR routing traffic to a gRPC service.
```

.Procedure

. Delete your `HTTPRoute` CRs. The `HTTPRoute` CR must be removed before the `GRPCRoute` CR can be used on the same hostname.

> . Update the `spec.targetRef.kind` parameter value in your `AuthPolicy` and `RateLimitPolicy` CRs to `GRPCRoute` by using the following example:

Reviewer 2: If you considered the above change, maybe consider this one: 

```suggestion
. Update the `spec.targetRef.kind` parameter value in your `AuthPolicy` and `RateLimitPolicy` CRs to `GRPCRoute` as in the following example:
```

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: PROCEDURE
[id="con-migrate-policies-to-grpc_{context}"]
= Migrating policies to `GRPCRoute` support

:FeatureName: `GRPCRoute` policy attachment support

include::snippets/technology-preview.adoc[]

[role="_abstract"]
When you want to replace your `HTTPRoute` custom resources (CRs) with `GRPCRoute` CRs, you must update your policies to reference a `GRPCRoute` CR, and configure other parameters in those policies as required for your use case.

.Prerequisites

* You installed {ocp}.
* You installed {oc-first}.
* You installed the {mcpg} Operator.
* You created a Gateway object.
* You created a `GRPCRoute` CR.

.Procedure

. Update the `spec.targetRef.kind` parameter value in your `AuthPolicy` and `RateLimitPolicy` CRs to `GRPCRoute` by using the following example:
+
.Example `AuthPolicy` CR updated to gRPC
[source,yaml]
----
#...
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: GRPCRoute
    name: grpcstore
#...
----

> . Delete your outdated `HTTPRoute` CRs.

Reviewer: The migration procedure assumes that a GRPCRoute can coexist alongside an existing HTTPRoute on the same hostname, but per the Gateway API spec this is not supported — the gateway implementation should reject the second route with Accepted=False. The procedure needs to account for this ordering constraint (the HTTPRoute must be removed before the GRPCRoute can be accepted on the same hostname), or recommend separate hostnames. If both gRPC and HTTP traffic must share a hostname, the recommended approach is to use HTTPRoute only. See the Hostname Conflict Resolution Between Route Types (https://github.com/Kuadrant/architecture/blob/main/rfcs/0017-grpcroute-support.md#hostname-conflict-resolution-between-route-types) section of RFC 0017 for details.

//Hold PlanPolicy until SDK is at least TechPreview
//. Update `PlanPolicy` CRs that reference HTTP-specific paths to use the well-known attributes, `request.grpc.service` and `request.grpc.method`.
//+
//.Example YAML block with gRPC traffic only
//[source,yaml]
//----
//#...
//when:
//  - predicate: request.grpc.service == 'UserService'
//  - predicate: request.grpc.method == 'GetUser'
//#...
//----
//+
//.Example YAML block for mixed traffic
//[source,yaml]
//----
//#...
//when:
//  - predicate: request.headers['content-type'].startsWith('application/grpc')
//  - predicate: request.grpc.service == 'UserService'
//#...
//----
//+
//If your `PlanPolicy` CR targets a specific `GRPCRoute` CR and not a `Gateway` object, you do not need the protocol check since only gRPC traffic flows through gRPC routes.

> //. Update your `PlanPolicy` CRs as needed with changes to `method` names and other details specific to your use case.

Reviewer: Shall we remove this?
